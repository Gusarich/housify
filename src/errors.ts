import { SourceLocation } from './grammar/ast';

export class CompilationError extends Error {
    name: string = 'CompilationError';
    location: SourceLocation;

    constructor(message: string, location: SourceLocation) {
        super(message);
        this.location = location;
    }

    toString(): string {
        return `${this.location.interval.getLineAndColumnMessage()}\n${this.name}: ${this.message}`;
    }
}

export class InternalError extends CompilationError {
    name: string = 'InternalError';
}

export class ParseError extends CompilationError {
    name: string = 'ParseError';

    toString(): string {
        // errors from ohm-js already include the location
        // we might want to handle this differently in the future
        return this.message;
    }
}

export class ResolveError extends CompilationError {
    name: string = 'ResolveError';
}

export class GenerationError extends CompilationError {
    name: string = 'GenerationError';
}

export class EmulationError extends CompilationError {
    name: string = 'EmulationError';
}
