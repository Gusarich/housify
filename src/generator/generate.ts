import { ConstantEvaluationError, InternalError } from '../errors';
import {
    AstExpression,
    AstHandler,
    AstHouse,
    AstModule,
    AstStatement,
    getAstNodeById,
    SourceLocation,
} from '../grammar/ast';
import { Action, ActionKind } from '../housing/actions';
import { ConditionKind } from '../housing/conditions';
import { EventType } from '../housing/events';
import { StatCompareMode, StatChangeMode } from '../housing/util';
import { CompilerContext } from '../resolver/context';
import { resolveStructPath } from '../resolver/resolve';
import { evaluateConstantExpression } from './evaluate';

export class WriterContext {
    actions: Action[];
    globalStats: Map<string, number>;
    playerStats: Map<string, number>;
    usedStats: Set<string>;
    statsMapping: Map<string, string>;
    ctx: CompilerContext;

    constructor(ctx: CompilerContext) {
        this.actions = [];
        this.globalStats = new Map();
        this.playerStats = new Map();
        this.usedStats = new Set();
        this.statsMapping = new Map();
        this.ctx = ctx;
    }

    clone() {
        const wctx = new WriterContext(this.ctx);
        wctx.actions = [...this.actions];
        wctx.globalStats = new Map(this.globalStats);
        wctx.playerStats = new Map(this.playerStats);
        wctx.usedStats = new Set(this.usedStats);
        wctx.statsMapping = new Map(this.statsMapping);
        return wctx;
    }

    mapStatName(stat: string) {
        const mappedStat = this.statsMapping.get(stat);
        return mappedStat ?? '$' + stat;
    }
}

export function wrapAsPlayerStat(stat: string) {
    return '%stat.player/' + stat + '%';
}

export function wrapAsGlobalStat(stat: string) {
    return '%stat.global/' + stat + '%';
}

export function isStatReference(stat: string) {
    return stat.startsWith('%stat.player/') || stat.startsWith('%stat.global/');
}

export function isGlobalStatReference(stat: string) {
    return stat.startsWith('%stat.global/');
}

export function unwrapStatReference(stat: string, location?: SourceLocation) {
    if (!isStatReference(stat)) {
        throw new InternalError('Not a stat reference', location);
    }
    return stat.slice(13, -1);
}

let tempId = 0;

function getNextTempId() {
    return tempId++;
}

export function resetTempId() {
    tempId = 0;
}

export type CompiledModule = {
    kind: 'module';
    houses: CompiledHouse[];
};

export type CompiledHouse = {
    kind: 'house';
    name: string;
    globalStats: string[];
    playerStats: string[];
    handlers: CompiledHandler[];
};

export type CompiledHandler = {
    kind: 'handler';
    event: EventType;
    actions: Action[];
};

export function cloneCompiledPiece(
    piece: CompiledModule | CompiledHouse | CompiledHandler,
): CompiledModule | CompiledHouse | CompiledHandler {
    switch (piece.kind) {
        case 'module':
            return {
                ...piece,
                houses: piece.houses.map(cloneCompiledPiece) as CompiledHouse[],
            };
        case 'house':
            return {
                ...piece,
                globalStats: [...piece.globalStats],
                playerStats: [...piece.playerStats],
                handlers: piece.handlers.map(
                    cloneCompiledPiece,
                ) as CompiledHandler[],
            };
        case 'handler':
            return {
                ...piece,
                actions: [...piece.actions],
            };
    }
}

export function writeExpression(
    expression: AstExpression,
    ctx: CompilerContext,
    wctx: WriterContext,
): string {
    try {
        const result = evaluateConstantExpression(expression, ctx);
        const tempStat = '$$' + getNextTempId();
        wctx.usedStats.add(tempStat);
        wctx.actions.push({
            kind: ActionKind.CHANGE_PLAYER_STAT,
            mode: StatChangeMode.SET,
            stat: tempStat,
            value: result,
        });
        return wrapAsPlayerStat(tempStat);
    } catch (e) {
        if (e instanceof ConstantEvaluationError && e.fatal) {
            throw e;
        }
    }

    switch (expression.kind) {
        case 'integerLiteral': {
            const tempStat = '$$' + getNextTempId();
            wctx.usedStats.add(tempStat);
            wctx.actions.push({
                kind: ActionKind.CHANGE_PLAYER_STAT,
                mode: StatChangeMode.SET,
                stat: tempStat,
                value: expression.value,
            });
            return wrapAsPlayerStat(tempStat);
        }
        case 'booleanLiteral': {
            const tempStat = '$$' + getNextTempId();
            wctx.usedStats.add(tempStat);
            wctx.actions.push({
                kind: ActionKind.CHANGE_PLAYER_STAT,
                mode: StatChangeMode.SET,
                stat: tempStat,
                value: expression.value ? '1' : '0',
            });
            return wrapAsPlayerStat(tempStat);
        }
        case 'expressionId': {
            return wrapAsPlayerStat(wctx.mapStatName(expression.name.name));
        }
        case 'expressionField':
            if (expression.struct.kind === 'expressionId') {
                if (expression.struct.name.name === 'global') {
                    return wrapAsGlobalStat(expression.field.name);
                } else if (expression.struct.name.name === 'player') {
                    return wrapAsPlayerStat(expression.field.name);
                }
            }
            return wrapAsPlayerStat(
                resolveStructPath(expression, wctx).join('.'),
            );

        case 'expressionUnary': {
            const operand = writeExpression(expression.operand, ctx, wctx);
            const tempStat = '$$' + getNextTempId();
            wctx.usedStats.add(tempStat);

            switch (expression.op) {
                case '-': {
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.SET,
                        stat: tempStat,
                        value: '0',
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.DECREMENT,
                        stat: tempStat,
                        value: operand,
                    });
                    break;
                }
                case '!': {
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.SET,
                        stat: tempStat,
                        value: '1',
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.DECREMENT,
                        stat: tempStat,
                        value: operand,
                    });
                    break;
                }
            }
            return wrapAsPlayerStat(tempStat);
        }
        case 'expressionBinary': {
            const left = writeExpression(expression.left, ctx, wctx);
            const right = writeExpression(expression.right, ctx, wctx);
            const tempStat = '$$' + getNextTempId();
            wctx.usedStats.add(tempStat);

            switch (expression.op) {
                case '+':
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.SET,
                        stat: tempStat,
                        value: left,
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.INCREMENT,
                        stat: tempStat,
                        value: right,
                    });
                    break;
                case '-':
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.SET,
                        stat: tempStat,
                        value: left,
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.DECREMENT,
                        stat: tempStat,
                        value: right,
                    });
                    break;
                case '*':
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.SET,
                        stat: tempStat,
                        value: left,
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.MULTIPLY,
                        stat: tempStat,
                        value: right,
                    });
                    break;
                case '/':
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.SET,
                        stat: tempStat,
                        value: left,
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.DIVIDE,
                        stat: tempStat,
                        value: right,
                    });
                    break;
                case '&&':
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.SET,
                        stat: tempStat,
                        value: left,
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.MULTIPLY,
                        stat: tempStat,
                        value: right,
                    });
                    break;
                case '||': {
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.SET,
                        stat: tempStat,
                        value: left,
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.INCREMENT,
                        stat: tempStat,
                        value: right,
                    });
                    const tempStat2 = '$$' + getNextTempId();
                    wctx.usedStats.add(tempStat2);
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.SET,
                        stat: tempStat2,
                        value: left,
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.MULTIPLY,
                        stat: tempStat2,
                        value: right,
                    });
                    wctx.actions.push({
                        kind: ActionKind.CHANGE_PLAYER_STAT,
                        mode: StatChangeMode.DECREMENT,
                        stat: tempStat,
                        value: wrapAsPlayerStat(tempStat2),
                    });
                    break;
                }
                case '==':
                case '!=':
                case '<':
                case '<=':
                case '>':
                case '>=':
                    wctx.actions.push({
                        kind: ActionKind.CONDITIONAL,
                        conditions: [
                            {
                                kind: isGlobalStatReference(left)
                                    ? ConditionKind.GLOBAL_STAT
                                    : ConditionKind.PLAYER_STAT,
                                stat: unwrapStatReference(
                                    left,
                                    expression.left.source,
                                ),
                                mode:
                                    expression.op === '==' ||
                                    expression.op === '!='
                                        ? StatCompareMode.EQUAL
                                        : expression.op === '>'
                                          ? StatCompareMode.GREATER_THAN
                                          : expression.op === '>='
                                            ? StatCompareMode.GREATER_THAN_OR_EQUAL
                                            : expression.op === '<'
                                              ? StatCompareMode.LESS_THAN
                                              : StatCompareMode.LESS_THAN_OR_EQUAL,
                                value: right,
                            },
                        ],
                        matchAny: false,
                        then: [
                            {
                                kind: ActionKind.CHANGE_PLAYER_STAT,
                                stat: tempStat,
                                mode: StatChangeMode.SET,
                                value: expression.op === '!=' ? '0' : '1',
                            },
                        ],
                        else: [
                            {
                                kind: ActionKind.CHANGE_PLAYER_STAT,
                                stat: tempStat,
                                mode: StatChangeMode.SET,
                                value: expression.op === '!=' ? '1' : '0',
                            },
                        ],
                    });
            }
            return wrapAsPlayerStat(tempStat);
        }
        case 'expressionCall': {
            const func = ctx.getInlineFunction(expression.function.name);
            if (!func) {
                throw new InternalError(
                    `Function not found: ${expression.function.name}`,
                    expression.source,
                );
            }

            const callWctx = wctx.clone();

            const statsMapping = new Map(wctx.statsMapping);

            for (const [index, arg] of expression.arguments.entries()) {
                const argStat = '$$' + getNextTempId();
                const value = writeExpression(arg, ctx, callWctx);
                callWctx.usedStats.add(argStat);
                statsMapping.set(func.parameters[index]!.name, argStat);
                callWctx.actions.push({
                    kind: ActionKind.CHANGE_PLAYER_STAT,
                    mode: StatChangeMode.SET,
                    stat: argStat,
                    value,
                });
            }

            callWctx.statsMapping = statsMapping;

            const ast = getAstNodeById(func.astId);
            if (!ast || ast.kind !== 'function') {
                throw new InternalError(
                    `Invalid function AST with ID ${func.astId}: ${ast?.kind}`,
                    expression.source,
                );
            }

            const returnStat = '$$' + getNextTempId();
            callWctx.usedStats.add(returnStat);
            for (const statement of ast.statements) {
                writeStatement(statement, returnStat, ctx, callWctx);
            }

            wctx.actions = [...callWctx.actions];
            wctx.usedStats = new Set(callWctx.usedStats);

            return wrapAsPlayerStat(returnStat);
        }
    }
}

export function writeStatement(
    statement: AstStatement,
    returnStat: string | null,
    ctx: CompilerContext,
    wctx: WriterContext,
) {
    switch (statement.kind) {
        case 'statementLet': {
            const stat = '$' + statement.name.name;
            wctx.usedStats.add(stat);
            const value = writeExpression(statement.value, ctx, wctx);
            wctx.actions.push({
                kind: ActionKind.CHANGE_PLAYER_STAT,
                mode: StatChangeMode.SET,
                stat,
                value,
            });
            break;
        }
        case 'statementAssign':
        case 'statementAugmentedAssign': {
            const mode =
                statement.kind === 'statementAssign'
                    ? StatChangeMode.SET
                    : statement.op === '+'
                      ? StatChangeMode.INCREMENT
                      : statement.op === '-'
                        ? StatChangeMode.DECREMENT
                        : statement.op === '*'
                          ? StatChangeMode.MULTIPLY
                          : statement.op === '/'
                            ? StatChangeMode.DIVIDE
                            : StatChangeMode.SET;
            if (statement.lvalue.kind === 'expressionId') {
                wctx.actions.push({
                    kind: ActionKind.CHANGE_PLAYER_STAT,
                    mode,
                    stat: wctx.mapStatName(statement.lvalue.name.name),
                    value: writeExpression(statement.value, ctx, wctx),
                });
            } else if (statement.lvalue.kind === 'expressionField') {
                if (statement.lvalue.struct.kind === 'expressionId') {
                    if (statement.lvalue.struct.name.name === 'global') {
                        wctx.actions.push({
                            kind: ActionKind.CHANGE_GLOBAL_STAT,
                            mode,
                            stat: statement.lvalue.field.name,
                            value: writeExpression(statement.value, ctx, wctx),
                        });
                        break;
                    }
                    if (statement.lvalue.struct.name.name === 'player') {
                        wctx.actions.push({
                            kind: ActionKind.CHANGE_PLAYER_STAT,
                            mode,
                            stat: statement.lvalue.field.name,
                            value: writeExpression(statement.value, ctx, wctx),
                        });
                        break;
                    }
                }
                wctx.actions.push({
                    kind: ActionKind.CHANGE_PLAYER_STAT,
                    mode,
                    stat: resolveStructPath(statement.lvalue, wctx).join('.'),
                    value: writeExpression(statement.value, ctx, wctx),
                });
            } else {
                throw new InternalError('Invalid lvalue', statement.source);
            }
            break;
        }
        case 'statementExpression':
            writeExpression(statement.expression, ctx, wctx);
            break;
        case 'statementIf': {
            const condition = writeExpression(statement.condition, ctx, wctx);

            const thenWctx = wctx.clone();
            for (const thenStatement of statement.then) {
                writeStatement(thenStatement, returnStat, ctx, thenWctx);
            }
            wctx.usedStats = new Set(thenWctx.usedStats);

            const elseWctx = wctx.clone();
            if (statement.else) {
                for (const elseStatement of statement.else) {
                    writeStatement(elseStatement, returnStat, ctx, elseWctx);
                }
            }
            wctx.usedStats = new Set(elseWctx.usedStats);

            wctx.actions.push({
                kind: ActionKind.CONDITIONAL,
                conditions: [
                    {
                        kind: isGlobalStatReference(condition)
                            ? ConditionKind.GLOBAL_STAT
                            : ConditionKind.PLAYER_STAT,
                        stat: unwrapStatReference(
                            condition,
                            statement.condition.source,
                        ),
                        mode: StatCompareMode.EQUAL,
                        value: '1',
                    },
                ],
                matchAny: false,
                then: thenWctx.actions,
                else: elseWctx.actions,
            });
            break;
        }
        case 'statementReturn':
            if (returnStat && statement.expression) {
                wctx.actions.push({
                    kind: ActionKind.CHANGE_PLAYER_STAT,
                    mode: StatChangeMode.SET,
                    stat: returnStat,
                    value: writeExpression(statement.expression, ctx, wctx),
                });
            }
            break;
    }
}

export function writeHandler(
    ast: AstHandler,
    ctx: CompilerContext,
    wctx: WriterContext,
): CompiledHandler {
    if (!(ast.event.name in EventType)) {
        throw new InternalError(
            `Invalid event type: ${ast.event.name}`,
            ast.event.source,
        );
    }

    for (const item of ast.statements) {
        writeStatement(item, null, ctx, wctx);
    }

    // Set all temp stats to 0
    for (const stat of wctx.usedStats) {
        wctx.actions.push({
            kind: ActionKind.CHANGE_PLAYER_STAT,
            mode: StatChangeMode.SET,
            stat,
            value: '0',
        });
    }

    return {
        kind: 'handler',
        event: ast.event.name as EventType,
        actions: wctx.actions,
    };
}

export function writeHouse(ast: AstHouse, ctx: CompilerContext): CompiledHouse {
    const house = ctx.getHouse(ast.name.name);
    if (!house) {
        throw new InternalError(
            `House not found: ${ast.name.name}`,
            ast.name.source,
        );
    }

    const globalStats: Map<string, number> = new Map();
    house.globalStats.forEach((stat, index) => {
        globalStats.set(stat.name, index);
    });
    const playerStats: Map<string, number> = new Map();
    house.playerStats.forEach((stat, index) => {
        playerStats.set(stat.name, index);
    });
    const handlers: CompiledHandler[] = [];

    for (const item of ast.items) {
        if (item.kind === 'handler') {
            resetTempId();
            handlers.push(writeHandler(item, ctx, new WriterContext(ctx)));
        }
    }

    return {
        kind: 'house',
        name: ast.name.name,
        globalStats: [...globalStats.keys()].sort(
            (a, b) => globalStats.get(a)! - globalStats.get(b)!,
        ),
        playerStats: [...playerStats.keys()].sort(
            (a, b) => playerStats.get(a)! - playerStats.get(b)!,
        ),
        handlers,
    };
}

export function writeModule(
    ast: AstModule,
    ctx: CompilerContext,
): CompiledModule {
    const houses: CompiledHouse[] = [];

    for (const item of ast.items) {
        if (item.kind === 'house') {
            houses.push(writeHouse(item, ctx));
        }
    }

    return {
        kind: 'module',
        houses,
    };
}
