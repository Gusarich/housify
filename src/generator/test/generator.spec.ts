import { ConstantEvaluationError } from '../../errors';
import { resetNodeId } from '../../grammar/ast';
import { parse } from '../../grammar/grammar';
import { CompilerContext } from '../../resolver/context';
import { resolveModule } from '../../resolver/resolve';
import { loadCases } from '../../utils/loadCases';
import { writeModule } from '../generate';

describe('Generator', () => {
    beforeEach(() => {
        resetNodeId();
    });

    for (const r of loadCases(__dirname + '/cases/')) {
        it('should generate ' + r.name, () => {
            const moduleAst = parse(r.code);
            const ctx = new CompilerContext();
            resolveModule(moduleAst, ctx);
            const module = writeModule(moduleAst, ctx);
            expect(module).toMatchSnapshot();
        });
    }
});

describe('Constant Evaluator', () => {
    beforeEach(() => {
        resetNodeId();
    });

    for (const r of loadCases(__dirname + '/cases-failed/')) {
        it('should fail ' + r.name, () => {
            const moduleAst = parse(r.code);
            const ctx = new CompilerContext();
            resolveModule(moduleAst, ctx);
            try {
                writeModule(moduleAst, ctx);
            } catch (e) {
                expect(e).toBeInstanceOf(ConstantEvaluationError);
                expect(
                    (e as ConstantEvaluationError).toString(),
                ).toMatchSnapshot();
            }
        });
    }
});
