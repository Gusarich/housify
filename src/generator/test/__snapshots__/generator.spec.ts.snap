// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generator should generate 1 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "amount": "%global.counter%",
              "mode": "SET",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "INCREMENT",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$0%",
              "mode": "SET",
              "stat": "counter",
              "type": "CHANGE_GLOBAL_STAT",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "TestHouse",
      "playerStats": [
        "$$0",
      ],
    },
  ],
  "kind": "module",
}
`;

exports[`Generator should generate 2 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "joinCounter",
        "quitCounter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "amount": "%global.joinCounter%",
              "mode": "SET",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "INCREMENT",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$0%",
              "mode": "SET",
              "stat": "joinCounter",
              "type": "CHANGE_GLOBAL_STAT",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
        {
          "actions": [
            {
              "amount": "%global.quitCounter%",
              "mode": "SET",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "INCREMENT",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$1%",
              "mode": "SET",
              "stat": "quitCounter",
              "type": "CHANGE_GLOBAL_STAT",
            },
          ],
          "event": "QUIT",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "TestHouse",
      "playerStats": [
        "$$0",
        "$$1",
      ],
    },
  ],
  "kind": "module",
}
`;

exports[`Generator should generate 3 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "amount": "%global.counter%",
              "mode": "SET",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "INCREMENT",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$0%",
              "mode": "SET",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "2",
              "mode": "INCREMENT",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$1%",
              "mode": "SET",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "3",
              "mode": "INCREMENT",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$2%",
              "mode": "SET",
              "stat": "$$3",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "4",
              "mode": "INCREMENT",
              "stat": "$$3",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$3%",
              "mode": "SET",
              "stat": "counter",
              "type": "CHANGE_GLOBAL_STAT",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "TestHouse",
      "playerStats": [
        "$$0",
        "$$1",
        "$$2",
        "$$3",
      ],
    },
  ],
  "kind": "module",
}
`;

exports[`Generator should generate 4 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "amount": "%global.counter%",
              "mode": "SET",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "INCREMENT",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$0%",
              "mode": "SET",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "-1",
              "mode": "INCREMENT",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$1%",
              "mode": "SET",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "4",
              "mode": "INCREMENT",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$2%",
              "mode": "SET",
              "stat": "counter",
              "type": "CHANGE_GLOBAL_STAT",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "TestHouse",
      "playerStats": [
        "$$0",
        "$$1",
        "$$2",
      ],
    },
  ],
  "kind": "module",
}
`;

exports[`Generator should generate 5 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "amount": "%global.counter%",
              "mode": "SET",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "2",
              "mode": "MULTIPLY",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "3",
              "mode": "SET",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$0%",
              "mode": "INCREMENT",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$1%",
              "mode": "SET",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "DECREMENT",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$2%",
              "mode": "SET",
              "stat": "$$3",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "4",
              "mode": "INCREMENT",
              "stat": "$$3",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$3%",
              "mode": "SET",
              "stat": "$$4",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "5",
              "mode": "INCREMENT",
              "stat": "$$4",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$4%",
              "mode": "SET",
              "stat": "counter",
              "type": "CHANGE_GLOBAL_STAT",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "TestHouse",
      "playerStats": [
        "$$0",
        "$$1",
        "$$2",
        "$$3",
        "$$4",
      ],
    },
  ],
  "kind": "module",
}
`;

exports[`Generator should generate 6 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "amount": "%global.counter%",
              "mode": "SET",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "2",
              "mode": "MULTIPLY",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "3",
              "mode": "SET",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$0%",
              "mode": "INCREMENT",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$1%",
              "mode": "SET",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "DECREMENT",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$2%",
              "mode": "SET",
              "stat": "$$3",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "4",
              "mode": "INCREMENT",
              "stat": "$$3",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$3%",
              "mode": "SET",
              "stat": "$$4",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "5",
              "mode": "INCREMENT",
              "stat": "$$4",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$4%",
              "mode": "SET",
              "stat": "$x",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$x%",
              "mode": "SET",
              "stat": "$$5",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "2",
              "mode": "MULTIPLY",
              "stat": "$$5",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "SET",
              "stat": "$$6",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$5%",
              "mode": "INCREMENT",
              "stat": "$$6",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$6%",
              "mode": "SET",
              "stat": "counter",
              "type": "CHANGE_GLOBAL_STAT",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "TestHouse",
      "playerStats": [
        "$x",
        "$$0",
        "$$1",
        "$$2",
        "$$3",
        "$$4",
        "$$5",
        "$$6",
      ],
    },
  ],
  "kind": "module",
}
`;

exports[`Generator should generate 7 1`] = `
{
  "houses": [
    {
      "globalStats": [
        "counter",
      ],
      "handlers": [
        {
          "actions": [
            {
              "amount": "1",
              "mode": "SET",
              "stat": "$x",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "0",
              "mode": "SET",
              "stat": "$y",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "SET",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$y%",
              "mode": "DECREMENT",
              "stat": "$$0",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$x%",
              "mode": "SET",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$0%",
              "mode": "MULTIPLY",
              "stat": "$$1",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$1%",
              "mode": "SET",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "0",
              "mode": "MULTIPLY",
              "stat": "$$2",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$2%",
              "mode": "SET",
              "stat": "$$3",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "INCREMENT",
              "stat": "$$3",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$2%",
              "mode": "SET",
              "stat": "$$4",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "1",
              "mode": "MULTIPLY",
              "stat": "$$4",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "$$4",
              "mode": "DECREMENT",
              "stat": "$$3",
              "type": "CHANGE_PLAYER_STAT",
            },
            {
              "amount": "%player.$$3%",
              "mode": "SET",
              "stat": "$z",
              "type": "CHANGE_PLAYER_STAT",
            },
          ],
          "event": "JOIN",
          "kind": "handler",
        },
      ],
      "kind": "house",
      "name": "TestHouse",
      "playerStats": [
        "$x",
        "$y",
        "$z",
        "$$0",
        "$$1",
        "$$2",
        "$$3",
        "$$4",
      ],
    },
  ],
  "kind": "module",
}
`;
